go/vt/wrangler/reparent.go

```diff
@@ -226,3 +179,117 @@ func (wr *Wrangler) ReparentTablet(ctx context.Context, tabletAlias topo.TabletA
+	if shardInfo.MasterAlias != masterElectTabletAlias {
+		if !force {
+			return fmt.Errorf("master-elect tablet %v is not the shard master, use -force to proceed anyway", masterElectTabletAlias)
+		}
+		wr.logger.Warningf("master-elect tablet %v is not the shard master, proceeding anyway as -force was used", masterElectTabletAlias)
+	}
+	if _, ok := masterTabletMap[masterElectTabletAlias]; !ok || len(masterTabletMap) != 1 {
+		if !force {
+			return fmt.Errorf("master-elect tablet %v is not the only master in the shard, use -force to proceed anyway", masterElectTabletAlias)
+		}
+		wr.logger.Warningf("master-elect tablet %v is not the only master in the shard, proceeding anyway as -force was used", masterElectTabletAlias)
+	}
+
+	// Tell the new master to break its slaves, return its replication
+	// position, and add a row to the reparent_journal table.
+	rp, err := wr.TabletManagerClient().InitMaster(ctx, masterElectTabletInfo)
+	if err != nil {
+		return err
+	}
+
+	// Now tell the new master to insert the reparent_journal row,
+	// and tell everybody else to become a slave of the new master,
+	// and wait for the row in the reparent_journal table.
+	// We start all these in parallel, to handle the semi-sync
+	// case: for the master to be able to commit its row in the
+	// reparent_journal table, it needs connected slaves.
+	now := time.Now().UnixNano()
+	wgMaster := sync.WaitGroup{}
+	wgSlaves := sync.WaitGroup{}
+	rec := concurrency.AllErrorRecorder{}
+	var masterErr error
+	for alias, tabletInfo := range tabletMap {
+		if alias == masterElectTabletAlias {
+			wgMaster.Add(1)
+			go func(alias topo.TabletAlias, tabletInfo *topo.TabletInfo) {
+				defer wgMaster.Done()
+				masterErr = wr.TabletManagerClient().PopulateReparentJournal(ctx, tabletInfo, now, initShardMasterOperation, alias, rp)
+			}(alias, tabletInfo)
+		} else {
+			wgSlaves.Add(1)
+			go func(alias topo.TabletAlias, tabletInfo *topo.TabletInfo) {
+				defer wgSlaves.Done()
+				if err := wr.TabletManagerClient().InitSlave(ctx, tabletInfo, masterElectTabletAlias, rp, now); err != nil {
+					rec.RecordError(fmt.Errorf("Tablet %v InitSlave failed: %v", alias, err))
+				}
+			}(alias, tabletInfo)
+		}
+	}
+
+	// After the master is done, we can update the shard record
+	// (note with semi-sync, it also needs at least one slave is done)
+	wgMaster.Wait()
+	if masterErr != nil {
+		wgSlaves.Wait()
+		return fmt.Errorf("failed to PopulateReparentJournal on master: %v", masterErr)
+	}
+	if shardInfo.MasterAlias != masterElectTabletAlias {
+		shardInfo.MasterAlias = masterElectTabletAlias
+		if err := topo.UpdateShard(ctx, wr.ts, shardInfo); err != nil {
+			wgSlaves.Wait()
+			return fmt.Errorf("failed to update shard master record: %v", err)
+		}
+	}
+
+	// Wait for the slaves to complete. If some of them fail, we
+	// don't want to rebuild the shard serving graph (the failure
+	// will most likely be a timeout, and our context will be
```

Comment:
If some of them fail quickly, then would you want to go ahead with the rebuild? If so, then I don't think we need to give that up here. If it is indeed because of a timeout, then the rebuild will fail as you said. If it's not a timeout, we could do something useful by letting the ones that succeeded start serving.